### PR TITLE
Add marker identifying which pickley performed the bootstrap

### PR DESCRIPTION
**Pull request by zsimic**
- Drop a marker in `.pk/.manifest/.bootstrap.json` identifying which pickley version performed the bootstrap
- Clean-up some key files on bootstrap, to avoid older pickley's remaining stuck, incapable of self-upgrading
- Allow main commands (install, upgrade, etc) to operate only on a bootstrapped base, secondary commands such as `describe` can work from anywhere (no "base" needed)
- Removed legacy code (pickley 4.3 not supported by bootstrap anymore)
- Better formalized "dev mode" (where `.venv/bin/pickley` operates on an implicit base in `.venv/dev_mode/`)
- More tests

### CODE REVIEW

File: src/pickley/__init__.py [hunk 1]
```diff
@@ -117,7 +117,7 @@ def update_from_pip_show(self, venv, canonical_name):
         self.canonical_name = canonical_name
         self.values = {}
         r = venv.run_pip("show", canonical_name, fatal=False)
-        if r.failed:
+        if r.failed:  # pragma: no cover, hard to trigger without excessive mocking
```
Comment (thatch) [inline src/pickley/__init__.py:120]: Is this really "excessive"?  My preference would be to leave paths that users could actually hit uncovered to remind me to cover them someday :)

```
def test():
  result = Mock(failed=True, full_output="foo")
  venv = Mock(run_pip=Mock(side_effect=result))
  pm = PipMetadata()
  pm.update_from_pip_show(venv, "cn")
  assert pm.problem == "Failed to `pip show cn`: foo"
```

File: src/pickley/__init__.py [hunk 2]
```diff
@@ -410,9 +410,6 @@ def __init__(self, given_package_spec: str, authoritative=False, settings=None):
     def __repr__(self):
         return runez.short(self.auto_upgrade_spec)
 
-    def __lt__(self, other):
```
Comment (thatch) [inline src/pickley/__init__.py:413]: You don't want it to be orderable-ish any more?

File: src/pickley/__init__.py [hunk 3]
```diff
@@ -665,6 +666,16 @@ def program_version(path, logger=None):
             if r.succeeded:
                 return PickleyConfig.parsed_version(r.output or r.error)
 
+    @staticmethod
+    def installed_sibling_folders(canonical_name):
+        """Sibling installations of the form '<canonical_name>-<version>'."""
+        regex = re.compile(r"^(.+)-(\d+[.\d+]+)$")
```
Comment (thatch) [inline src/pickley/__init__.py:672]: This doesn't handle any pre or post releases, is there a reason for it to be overly strict?

Is there a reason to allow `+` in the character class, are you trying to support local identifiers?  If you intend to allow them, they are not restricted to digits.

[pypi-simple uses](https://github.com/jwodder/pypi-simple/blob/a357d83ad83ac54a81c0abb6097b9858e424f222/src/pypi_simple/filenames.py#L9) uses `VERSION_NODASH` which is short and overly permissive.

File: src/pickley/__init__.py [hunk 4]
```diff
@@ -665,6 +666,16 @@ def program_version(path, logger=None):
...
+        for item in runez.ls_dir(CFG.meta):
+            if item.is_dir():
+                m = regex.match(item.name)
```
Comment (thatch) [inline src/pickley/__init__.py:675]: you can use `.fullmatch` here and you don't have to wrap with `^...$` (it's safer too, more like `^(?:...)$` which is a little clunky to read

File: src/pickley/bstrap.py [hunk 5]
```diff
@@ -135,7 +134,7 @@ def bootstrap_pickley_with_pip(self, venv_folder: Path):
         if not needs_virtualenv and not DRYRUN:
             needs_virtualenv = not is_executable(pip)
 
-        if needs_virtualenv:
+        if needs_virtualenv:  # pragma: no cover, not testing py3.6 fallback anymore
```
Comment (thatch) [inline src/pickley/bstrap.py:137]: `requires_python` seems to indicate you can work on 3.6, is that not true?

File: src/pickley/__init__.py (hunk 2 above)
Comment (zsimic) [reply src/pickley/__init__.py:413]: This was a left-over from a long time ago where I used to sort `PackageSpec`-s, now the package specs are:
- always in the same order as given on CLI (ie: `pickley install foo bar` -> the order `foo, bar` is respected as-is)
- finding all installed is done by sorting on name, like `[PackageSpec(x) for x in sorted(installed_names)]`

So, sorting by `PackageSpec` is not used (and the test coverage was showing that).

File: src/pickley/__init__.py (hunk 3 above)
Comment (zsimic) [reply src/pickley/__init__.py:672]: Ouch! Good point, this old code needs to be adapted! As-is, it would never clean up a `1.0rc2` for example...

File: src/pickley/cli.py [hunk 6]
```diff
@@ -517,7 +546,7 @@ def cmd_list(border, format, verbose):
         python = runez.dim("-not needed-") if pspec.is_uv else manifest and manifest.python_executable
         delivery = runez.dim("-") if pspec.is_uv else manifest and manifest.delivery
         package_manager = manifest and manifest.package_manager
-        if package_manager and package_manager != default_package_manager:
+        if package_manager and package_manager != default_package_manager:  # pragma: no cover, uncommon
```
Comment (thatch) [inline src/pickley/cli.py:549]: I'm not a fan of this no-cover either, just take the not-100%-coverage or write a test

**Review by thatch — CHANGES_REQUESTED** [review]
On the right path, but this is a lot of changes in one PR.

File: src/pickley/bstrap.py (hunk 5 above)
Comment (zsimic) [reply src/pickley/bstrap.py:137]: Yes, here I would need to simulate a `python3 -mvenv ...` failing. Some distros for example do not provide the std lib module `venv`. Some very old python versions have a buggy `venv` module (IIRC).

Here, the intent is to double check on whether `-mvenv` actually did its job, and if it didn't, then download and call `python .../virtualenv.pyz ...` as a fallback.

This is known to work right now, but yeah wonder if I should keep testing this (as it requires a complex test setup), since this is not a thing anymore now for ages (`-mvenv` failing) + we don't even need `-mvenv` with uv.

File: src/pickley/cli.py (hunk 6 above)
Comment (zsimic) [reply src/pickley/cli.py:549]: It's a hard balance to strike, I'm struggling with it :)
I use code coverage to find what's the next edge case I'm missing a test for, and mark with `no cover` the ones I think are not worth setting up a complex test for.

Later, every now and then, I search for `no cover`, remove it, and see if I could add a test now (this is here to stay, and an even a complex test would be warranted)

Comment (zsimic) [reply src/pickley/cli.py:549]: Also, after a refactor, I typically remove all `no cover` pragmas, comment out all the complex not-so-useful mocks, and see which part of the code is now ripe for removal entirely (ie: code that remained "covered" only because there is an esoteric test that exercised it)
